pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -510,25 +524,120 @@ private CompletableFuture<Integer> closeServiceUnit(String serviceUnit) {
-                        NamespaceBundleSplitAlgorithm.of(pulsar.getConfig().getDefaultNamespaceBundleSplitAlgorithm()),
-                        null)
-                .whenComplete((__, ex) -> {
-                    double splitBundleTime = TimeUnit.NANOSECONDS
-                            .toMillis((System.nanoTime() - startTime));
-                    if (ex == null) {
-                        log.info("Successfully split {} namespace-bundle in {} ms",
-                                serviceUnit, splitBundleTime);
-                    } else {
-                        log.error("Failed to split {} namespace-bundle in {} ms",
-                                serviceUnit, splitBundleTime, ex);
-                    }
-                });
+        NamespaceService namespaceService = pulsar.getNamespaceService();
+        NamespaceBundleFactory bundleFactory = namespaceService.getNamespaceBundleFactory();
+        NamespaceBundle bundle = getNamespaceBundle(serviceUnit);
+        CompletableFuture<Void> completionFuture = new CompletableFuture<>();
+        final AtomicInteger counter = new AtomicInteger(0);
+        this.splitServiceUnitOnceAndRetry(namespaceService, bundleFactory, bundle, serviceUnit, data,
+                counter, startTime, completionFuture);
+        return completionFuture;
+    }
+
+    @VisibleForTesting
+    protected void splitServiceUnitOnceAndRetry(NamespaceService namespaceService,
+                                                NamespaceBundleFactory bundleFactory,
+                                                NamespaceBundle bundle,
+                                                String serviceUnit,
+                                                ServiceUnitStateData data,
+                                                AtomicInteger counter,
+                                                long startTime,
+                                                CompletableFuture<Void> completionFuture) {
+        CompletableFuture<List<NamespaceBundle>> updateFuture = new CompletableFuture<>();
+
+        getSplitBoundary(bundle).thenAccept(splitBundles -> {
+            // Split and updateNamespaceBundles. Update may fail because of concurrent write to Zookeeper.
+            if (splitBundles == null) {
+                String msg = format("Bundle %s not found under namespace", serviceUnit);
+                updateFuture.completeExceptionally(new BrokerServiceException.ServiceUnitNotReadyException(msg));
+                return;
+            }
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker());
+            for (NamespaceBundle sBundle : splitBundles.getRight()) {
+                futures.add(pubAsync(sBundle.toString(), next).thenAccept(__ -> {}));
+            }
+            NamespaceName nsname = bundle.getNamespaceObject();
+            FutureUtil.waitForAll(futures).thenRun(() ->
```

Comment:
Should we wait til getOwner(newBundles) == this.broker?

Reply:
I have been thinking, should we maintain a map instead of a table view in `ServiceUnitStateChannelImpl`, so we can put the latest value to the local map when sending success the state message to the topic?

Reply:
I am not sure why we need a separate map here.
We can make a blocking call until getOwner(newBundles).join(timeout ).get() == this.broker -- this will ensure that this broker's table-view owns the new bundles for the lookup requests. 

 If timeout, I assume we can retry on this, and this retry should be idempotent.

Reply:
If we make the new children's bundle status as `assigned`, then we can make sure get the right result when doing a lookup request, right? Of course, it needs a compaction strategy.

Reply:
Yes, `Owned -> Assigned` is an invalid transition(without source broker,  non-transfer), so the later `Assigned` will be ignored (conflict resolution).

Reply:
So if we set the bundle status as `Assigned`, then we don't need to block the call until `getOwner(newBundles).join(timeout ).get() == this.broker`.

Reply:
Sorry, I think we need to revert this next back to `ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker());`

`Free->Assigned(Non-Transfer)` is for the state when broker A trying to assign bundle ownership to broker B when there is no ownership, `Free`. (Since broker A assumes the broker B is online, `Assigned-> Owned` change requires broker B to to acknowledge `Assigned` by `Owned` broadcast. 

As we defined in the `ServiceUnitState` state diagram, I think the children's state can be `Owned` initially (does not need to be `Assigned`, as the parent broker(the same broker) always owns the children).

My initial concern was possible racing conditions if the broker does not wait until it receives its `Owned` message in its table-view, before updating the bundle partitions in ZK(case 2 below). I think this waiting seems redundant but still safe to have it. 

Let's review the racing conditions here.

case 1: when lookup parent(before updating zk bundle partitions) bundle before/after the broker receives children `Owned` in its table-view
=> happy case: the broker immediately returns the parent bundle lookup result(the same broker).

case 2: when lookup children(after updating zk bundle partitions) bundles before the broker receives children `Owned` in its table-view
=> unhappy case: the broker runs the bundle assignment logic and probably tries to assign the ownership to another broker. However, this `Owned->Assigned(Non-Transfer)` transition will be filtered out as invalid. Eventually, the children bundle lookup will either timeout or correctly return the original owner broker when the first `Owned` msg arrives. Still, it will be safe to avoid this case if we can delay zk bundle partition updates til `getOwner(children bundles).join(timeout ).get() == this.broker.` 

case 3: when lookup children(after updating zk bundle partitions) after the broker receives children `Owned` in its table-view
=> happy case: the broker immediately returns the lookup result.

Reply:
How about we treat the state as `Assigned(Transfer)` -> `new ServiceUnitStateData(Assigned, data.broker(), data.broker());`? 

I am unsure why we need to wait `getOwner(children bundles).join(timeout ).get() == this.broker`, because the wait can only ensure the current broker state, but if the client does lookup to another broker, it will still have a chance to assign the ownership to another broker, since the broker not received the `Owned ` message yet. If I understand it correctly.

Reply:
> I am unsure why we need to wait getOwner(children bundles).join(timeout ).get() == this.broker, because the wait can only ensure the current broker state, but if the client does lookup to another broker, it will still have a chance to assign the ownership to another broker, since the broker not received the Owned message yet. If I understand it correctly.

Your understanding is correct here. `getOwner(children bundles).join(timeout ).get() == this.broker,` cannot fully prevent the global racing condition here.

Even with `new ServiceUnitStateData(Assigned, data.broker(), data.broker())`,  some of the brokers might not see this `Assigned` upon child bundle lookup requests.

We should rely on the conflict resolution here to fully resolve this race condition.

So, I am fine with the original below `next` without waiting, `getOwner(children bundles).join(timeout ).get() == this.broker`

`ServiceUnitStateData next = new ServiceUnitStateData(Owned, data.broker());`

Reply:
Sure, in this case, using `Owned` will be better. We can use `Assigned` when splitting the bundle to a specific broker. I will push other PR to implement it.